aggregation/src/lib.rs

```diff
@@ -52,45 +53,309 @@ alloy_sol_types::sol! {
+    pub fn initial(self_image_id: impl Into<Digest>) -> Self {
+        Self {
+            self_image_id: self_image_id.into(),
+            mmr: MerkleMountainRange::empty(),
+        }
+    }
+
+    /// Returns true is this is the initial state, for an empty claim set.
+    pub fn is_initial(&self) -> bool {
+        self.mmr.is_empty()
+    }
+
+    /// Encodes the [GuestState] for committing to the journal. Uses a specialized codec.
+    /// See [MerkleMountainRange::encode].
+    pub fn encode(&self) -> Vec<u8> {
+        [self.self_image_id.as_bytes(), &self.mmr.encode()].concat()
+    }
+
+    /// Decodes the [GuestState] for the journal. Uses a specialized codec.
+    /// See [MerkleMountainRange::encode].
+    pub fn decode(bytes: impl AsRef<[u8]>) -> Result<Self, DecodingError> {
+        // Read the first 32 bytes as the self_image_id.
+        let (chunk, bytes) = bytes
+            .as_ref()
+            .split_at_checked(U256::BYTES)
+            .ok_or(DecodingError::UnexpectedEnd)?;
+        let self_image_id = Digest::try_from(chunk).unwrap();
+        let mmr = MerkleMountainRange::decode(bytes)?;
+        Ok(Self { self_image_id, mmr })
+    }
+
+    /// Create a [GuestInput] from this [GuestState]. When run with the guest, the given claims
+    /// will be accumulated into the Merkle mountain range, and will be finalized if `finalize` is
+    /// set to `true`.
+    ///
+    /// Will return an error if the [MerkleMountainRange] on the [GuestState] is already
+    /// finalized, as no more claims may be added and the guest would reject this input.
+    pub fn into_input(
+        self,
+        claims: Vec<ReceiptClaim>,
+        finalize: bool,
+    ) -> Result<GuestInput, Error> {
+        if self.mmr.is_finalized() {
+            return Err(Error::FinalizedError);
+        }
+        Ok(GuestInput {
+            state: self,
+            claims,
+            finalize,
+        })
+    }
+}
+
+/// Incrementally constructable Merkle mountain range.
+///
+/// Each entry in the list is a pair of (digest, max-depth), where max-depth tracks an upper bound
+/// on the size of the subtree for which the digest is the root. The largest subtree is at index 0,
+/// the smallest at index len - 1.
+///
+/// Note that the max size of the internal vec of digests (peaks) is equal to log_2 n where n is
+/// the number of leaves in the tree.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+pub struct MerkleMountainRange(Vec<Peak>);
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+struct Peak {
+    /// Digest for the root of the Merkle subtree committed to by this peak.
+    digest: Digest,
+    /// An upper-bound on the depth of the subtree rooted rooted at this peak.
+    ///
+    /// It is expressed as the total height of the subtree - 1, such that a peak with a single node
+    /// (i.e. a leaf) has a max_depth value of 0.
+    max_depth: u8,
```

Comment:
We should mention 0xff for a finalized MMR here

Reply:
6c34c00